sklearn/decomposition/_fastica.py

```diff
@@ -551,9 +575,27 @@ def g(x, fun_args):
             XT -= X_mean[:, np.newaxis]
 
             # Whitening and preprocessing by PCA
-            u, d, _ = linalg.svd(XT, full_matrices=False, check_finite=False)
+            if self.whiten_solver == "eigh":
+                # Faster when num_samples >> n_features
+                d, u = linalg.eigh(XT.dot(X))
+                sort_indices = np.argsort(d)[::-1]
+                eps = np.finfo(np.double).eps
+                degenerate_idx = d < eps
+                if np.any(degenerate_idx):
+                    warnings.warn(
+                        "There are some small singular values, using "
+                        "whiten_solver = 'svd' might lead to more "
+                        "accurate results."
+                    )
+                d[degenerate_idx] = eps  # For numerical issues
+                np.sqrt(d, out=d)
+                d, u = d[sort_indices], u[:, sort_indices]
+            else:
+                u, d = linalg.svd(XT, full_matrices=False, check_finite=False)[:2]
+
+            # Give consistent eigenvectors for both svd solvers
+            u *= np.sign(u[0])
```

Comment:
This PR is about adding a new eigenvector solver, `eigh` for whitening. The eigenvectors returned by `eigh` could be different than `svd` up to a sign flip. I see two options here:

1. Do a svd flip so that the solvers are consistent, but results in a "change model" for the SVD solver.
2. Do not do flipping, and document that the solvers can result in different models.

@glemaitre WDYT?

Reply:
For PCA, we do sign flipping for the SVD and randomized SVD. I would therefore go for consistency and make it a fix. Does it seem a bit too much to give control to this flipping through a parameter?

Reply:
> Does it seem a bit too much to give control to this flipping through a parameter?

This is feels like a major change to me, so I am okay with a parameter to control this.

Although, the sign flipping logic is a little different from [svd_flip](https://github.com/scikit-learn/scikit-learn/blob/b0a30983ead8de1825488522f9c8b699c30ec32e/sklearn/utils/extmath.py#L718). This implementation forces the first row to the positive, which works for all seeds in the `test_fastica_simple_different_solvers` test. We tried the sign flipping logic in [svd_flip](https://github.com/scikit-learn/scikit-learn/blob/b0a30983ead8de1825488522f9c8b699c30ec32e/sklearn/utils/extmath.py#L718), but it did not work for some seeds in `test_fastica_simple_different_solvers`.

@Micky774 Can you investigate why [svd_flip](https://github.com/scikit-learn/scikit-learn/blob/b0a30983ead8de1825488522f9c8b699c30ec32e/sklearn/utils/extmath.py#L718) like logic did not work? (I have a snippet here: https://github.com/scikit-learn/scikit-learn/pull/22527#discussion_r825507817)

Reply:
Will take a look at it, most likely early next week 👍

Reply:
Okay it looks like it fails a test where each column of `u` is populated by the same value give or take signs, e.g. 
```python
u = [[-0.70710678  0.70710678]
     [-0.70710678 -0.70710678]]
```
and in this case `np.argmax(np.abs(u), axis=0)` is a bit of a coin-toss regarding whether it selects the first or second element of each column, whereas normalizing so that the first column is positive is well-defined even in this case. Note that it's not necessary that all the elements of `u` are the same by magnitude, just that the elements of a given column are.

Reply:
So we could either:

1. Modify the test to account for this behavior if we want to keep `svd_flip` semantics.
2. Accept the new, admittedly arbitrary, sign semantics. 

For modifying the test, I'm not sure if there's really a clean way to do it, since we're testing the final output which does get pretty messed up even by this small difference.